plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -405,37 +407,163 @@ public void downloadPlugins(List<Plugin> plugins) {
      * @return set of all requested plugins and their recursive dependencies
      */
     public Map<String, Plugin> findPluginsAndDependencies(List<Plugin> requestedPlugins) {
-        Map<String, Plugin> allPluginDependencies = new HashMap<>();
-
-        for (Plugin requestedPlugin : requestedPlugins) {
-            //for each requested plugin, find all the dependent plugins that will be downloaded (including requested plugin)
-            Map<String, Plugin> dependencies = resolveRecursiveDependencies(requestedPlugin);
-
-            for (Plugin dependentPlugin : dependencies.values()) {
-                String dependencyName = dependentPlugin.getName();
-                VersionNumber dependencyVersion = dependentPlugin.getVersion();
-                if (!allPluginDependencies.containsKey(dependencyName)) {
-                    allPluginDependencies.put(dependencyName, dependentPlugin);
-                } else {
-                    Plugin existingDependency = allPluginDependencies.get(dependencyName);
-                    if (existingDependency.getVersion().compareTo(dependencyVersion) < 0) {
-                        logVerbose(String.format(
-                                "Version of %s (%s) required by %s (%s) is lower than the version required (%s) " +
-                                        "by %s (%s), upgrading required plugin version",
-                                dependencyName,
-                                existingDependency.getVersion().toString(),
-                                existingDependency.getParent().getName(),
-                                existingDependency.getParent().getVersion().toString(),
-                                dependencyVersion.toString(),
-                                dependentPlugin.getParent().getName(),
-                                dependentPlugin.getParent().getVersion().toString()));
-
-                        allPluginDependencies.replace(existingDependency.getName(), dependentPlugin);
+        requestedPlugins.parallelStream().forEach(this::resolveRecursiveDependencies);
+
+        Map<String, List<Plugin>> allPlugins = new HashMap<>();
+        for (Plugin p : requestedPlugins) {
+            Map<String, Plugin> optimizedChildren = findOptimalDependencies(p);
+            addToAllDependentPlugins(optimizedChildren, allPlugins);
+        }
+
+        Map<String, Plugin> optimizedPlugins = optimize(allPlugins);
+        return optimizedPlugins;
+    }
+
+    /**
+     * For a given plugin, recursively finds all "optimal" dependencies (i.e. repeated plugins are replaced with the
+     * highest required version
+     *
+     * @param plugin plugin to find dependencies
+     * @return map of <PluginName, Plugin> pairs corresponding to all optimal dependencies for the plugin passed in as
+     * a parameter
+     */
+    public Map<String, Plugin> findOptimalDependencies(Plugin plugin) {
```

Comment:
would you be able to time a non trivial plugin set before and after this change?

Reply:
Sure. I will do it on the list of plugins Joseph sent before: https://gist.github.com/casz/92b43e9ba4f994114b01d6ae9dc73b10

Reply:
So using that set up plugins up until xunit -
Downloading the plugins
Currently on master: 10.45 user 6.88 system 2:30.40 elapsed
After this change: 10.89 user 7.24 system 2:34.87 elapsed

Listing plugins (which will do the dependency resolution)
Currently on master: 3.65user 0.38system 0:13.55elapsed
After this change: 3.63user 0.38system 0:13.33elapsed

So it looks like these changes don't really have much of a performance impact - the bottleneck seems to be downloading them.

Reply:
How many times did you run it btw? normally want to run it a few times to get an average

Reply:
Not really sure how to take CPU usage differences into account, but I ran it four times on each branch.

Master --list
3.60user 0.40system 0:19.99elapsed 20%CPU
3.64user 0.37system 0:15.03elapsed 26%CPU
3.69user 0.39system 0:14.66elapsed 27%CPU
3.56user 0.36system 0:13.08elapsed 30%CPU
Average: 0:15.69

Master  download
10.18user 6.61system 2:38.13elapsed 10%CPU
10.18user 6.36system 3:10.54elapsed 8%CPU
10.38user 6.64system 2:48.37elapsed 10%CPU (
10.37user 6.70system 3:10.41elapsed 8%CPU
Average: 176.86s

This branch --list
3.70user 0.37system 0:15.01elapsed 27%CPU
3.62user 0.37system 0:13.82elapsed 28%CPU
3.71user 0.38system 0:16.91elapsed 24%CPU
3.61user 0.39system 0:17.13elapsed 23%CPU
Average: 0:15.71

This branch download: 
10.76user 7.08system 3:15.37elapsed 9%CPU
11.12user 7.77system 3:28.61elapsed 9%CPU
10.91user 7.82system 4:13.34elapsed 7%CPU
10.75user 7.14system 3:07.16elapsed 9%CPU
Average: 196.12s

Reply:
Looks pretty similar with one outlier probably due to network...